File: windows/ScoringSystem/Models/FoulType.cs
using System.Text.Json.Serialization;

namespace ScoringSystem.Models;

public class FoulType
{
    [JsonPropertyName("id")]
    public string Id { get; set; } = string.Empty;

    [JsonPropertyName("sportType")]
    public string SportType { get; set; } = string.Empty;

    [JsonPropertyName("code")]
    public string Code { get; set; } = string.Empty;

    [JsonPropertyName("name")]
    public string Name { get; set; } = string.Empty;

    [JsonPropertyName("nameEn")]
    public string NameEn { get; set; } = string.Empty;

    [JsonPropertyName("severity")]
    public string Severity { get; set; } = "common";

    [JsonPropertyName("penalty")]
    public FoulPenalty Penalty { get; set; } = new();

    [JsonPropertyName("description")]
    public string Description { get; set; } = string.Empty;
}

public class FoulPenalty
{
    [JsonPropertyName("type")]
    public string Type { get; set; } = string.Empty;

    [JsonPropertyName("duration")]
    public int? Duration { get; set; }

    [JsonPropertyName("points")]
    public int? Points { get; set; }

    [JsonPropertyName("description")]
    public string Description { get; set; } = string.Empty;
}

public class FoulTypeResponse
{
    [JsonPropertyName("success")]
    public bool Success { get; set; }

    [JsonPropertyName("data")]
    public List<FoulType> Data { get; set; } = new();

    [JsonPropertyName("total")]
    public int Total { get; set; }

    [JsonPropertyName("sportTypes")]
    public List<string> SportTypes { get; set; } = new();
}

public class SportFoulInfo
{
    [JsonPropertyName("name")]
    public string Name { get; set; } = string.Empty;

    [JsonPropertyName("nameEn")]
    public string NameEn { get; set; } = string.Empty;

    [JsonPropertyName("emoji")]
    public string Emoji { get; set; } = string.Empty;

    [JsonPropertyName("foulCount")]
    public int FoulCount { get; set; }
}

public class SportInfoResponse
{
    [JsonPropertyName("success")]
    public bool Success { get; set; }

    [JsonPropertyName("data")]
    public Dictionary<string, SportFoulInfo> Data { get; set; } = new();
}

File: android/RefereeApp/app/src/main/java/com/eventcontrol/refereeapp/domain/model/FoulType.kt
package com.eventcontrol.refereeapp.domain.model

import androidx.compose.ui.graphics.Color

/**
 * 犯规类型数据模型 - 从后端 API /api/foul-types 获取
 */
data class FoulType(
    val id: String,
    val sportType: String,
    val code: String,
    val name: String,
    val nameEn: String,
    val severity: FoulSeverity,
    val penalty: FoulPenalty,
    val description: String
)

/**
 * 犯规严重程度
 */
enum class FoulSeverity(val displayName: String, val color: androidx.compose.ui.graphics.Color) {
    MINOR("轻微", androidx.compose.ui.graphics.Color(0xFF4CAF50)),      // Green
    COMMON("普通", androidx.compose.ui.graphics.Color(0xFFFF9800)),     // Orange
    SERIOUS("严重", androidx.compose.ui.graphics.Color(0xFFFF5722)),    // Deep Orange
    TECHNICAL("技术犯规", androidx.compose.ui.graphics.Color(0xFF9C27B0)), // Purple
    FLAGRANT("恶意犯规", androidx.compose.ui.graphics.Color(0xFFF44336)); // Red

    companion object {
        fun fromString(value: String): FoulSeverity {
            return when (value.lowercase()) {
                "minor" -> MINOR
                "common" -> COMMON
                "serious" -> SERIOUS
                "technical" -> TECHNICAL
                "flagrant" -> FLAGRANT
                else -> COMMON
            }
        }
    }
}

/**
 * 犯规处罚方式
 */
data class FoulPenalty(
    val type: String,           // possession, free_throw, penalty_shot, card, etc.
    val duration: Int?,         // 处罚时长（秒），如罚下场时间
    val points: Int?,           // 罚分
    val description: String     // 处罚描述
) {
    fun getDisplayText(): String {
        return when (type) {
            "possession" -> "对方球权"
            "free_throw" -> "罚球${points ?: ""}次"
            "penalty_shot" -> "点球"
            "card" -> "出示${description}"
            "timeout" -> "暂停"
            "ejection" -> "驱逐出场"
            else -> description.ifEmpty { "无处罚" }
        }
    }
}

/**
 * 犯规类型 API 响应
 */
data class FoulTypeResponse(
    val success: Boolean,
    val data: List<FoulTypeDto>,
    val total: Int,
    val sportTypes: List<String>
)

/**
 * 犯规类型 DTO（用于 API 解析）
 */
data class FoulTypeDto(
    val id: String,
    val sportType: String,
    val code: String,
    val name: String,
    val nameEn: String,
    val severity: String,
    val penalty: PenaltyDto,
    val description: String
)

data class PenaltyDto(
    val type: String,
    val duration: Int?,
    val points: Int?,
    val description: String
)

/**
 * 扩展函数：将 DTO 转换为领域模型
 */
fun FoulTypeDto.toDomain(): FoulType {
    return FoulType(
        id = id,
        sportType = sportType,
        code = code,
        name = name,
        nameEn = nameEn,
        severity = FoulSeverity.fromString(severity),
        penalty = FoulPenalty(
            type = penalty.type,
            duration = penalty.duration,
            points = penalty.points,
            description = penalty.description
        ),
        description = description
    )
}
